raft-tech/TANF-app
# Feature/1351 - elastic

## Summary of Changes
Pull request closes #1351 

Adds
- Elastic search development infrastructure in docker compose
- Integration of officially maintained `django-elasticsearch-dsl` library for elastic documents, as well as a Django Rest Framework wrapper (`django-elasticsearch-dsl-drf`)
- Set up Django models and elastic search mappings for automatic index creation/updates

The `raft` space in Cloud.Gov has been set up with an elasticsearch instance. The following will still need to be done for other environments
- Create an `elasticsearch` service in Cloud.Gov
- Bind the service instance to the associated backend application
- Add three environment variables to the application, the values of which can be found in the logs after binding the service
   - `ELASTIC_HOST`
   - `AWS_ELASTIC_ACCESS_KEY`
   - `AWS_ELASTIC_SECRET`

## How to Test

*Note*: Ensure the `ELASTIC_HOST` environment variable is set in your `.env` file, you can use the value provided in `.env.example`

```
cd tdrs-frontend && npm run start
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in as an account with superuser access
1. Navigate to the Django admin and create a new `T1`, `T2`, etc. document under the `Search indexes` header
1. Verify (*) document updates happen in elastic search
   - (*) there's currently no way to verify in deployed environments. if the elastic connection is set up incorrectly, the creation of the model via the Django admin will fail, but we need to set up a Kibana proxy in order to view the data submitted to the elastic instance.
   - Locally, you can make a `GET` request to `http://localhost:9200/{index_name}/_search` and see the new data added
      - e.g. `GET http://localhost:9200/t6_submissions/_search`

https://user-images.githubusercontent.com/111305129/193842484-923fbaf2-7950-4a90-990b-5fdd0eb6d985.mov

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Each section / transmission format has an associated [index template](https://opensearch.org/docs/opensearch/index-templates/)
+ [x] We have a way to automatically configure Elasticsearch with these templates

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested? **(see comment in `search_indexes/tests.py`)**
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested? **N/A**
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] If this PR introduces dependencies, are their licenses documented? **Let's discuss**

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
